### ISSUE TITLE
Feature Request: Support for Ministral-8B-Instruct-2410

### ISSUE THREAD

**Issue by arch-btw** [ggerganov/llama.cpp#9914]
### Prerequisites

- [X] I am running the latest code. Mention the version if possible as well.
- [X] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [X] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [X] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

Dear llama.ccp team,

I would like to kindly request support for ```Ministral-8B-Instruct-2410```, if possible.

### Motivation

The motivation for this is that it is a new major Mistral release, and I think it will be a good addition for ```llama.cpp``` users.
Here is some information about the model from the HuggingFace page:

-------------

**Ministral 8B Key features**

- Released under the Mistral Research License, reach out to us for a commercial license
- Trained with a 128k context window with interleaved sliding-window attention
- Trained on a large proportion of multilingual and code data
- Supports function calling
- Vocabulary size of 131k, using the V3-Tekken tokenizer

-------------

Ministral 8B Architecture

| Feature               | Value                |
|:---------------------:|:--------------------:|
| **Architecture**      | Dense Transformer    |
| **Parameters**        | 8,019,808,256        |
| **Layers**            | 36                   |
| **Heads**             | 32                   |
| **Dim**               | 4096                 |
| **KV Heads (GQA)**    | 8                    |
| **Hidden Dim**        | 12288                |
| **Head Dim**          | 128                  |
| **Vocab Size**        | 131,072              |
| **Context Length**    | 128k                 |
| **Attention Pattern** | Ragged (128k,32k,32k,32k) |

### Possible Implementation

Release page: https://mistral.ai/news/ministraux/

HuggingFace page: https://huggingface.co/mistralai/Ministral-8B-Instruct-2410

Thank you for your time and consideration.

**Comment by ddh0** [comment]
Commits [6428ccf](https://github.com/mistralai/mistral-inference/commit/6428ccf99e4fa6acdb0176d5c8d77b2878c75040) and [fffa5da](https://github.com/mistralai/mistral-inference/commit/fffa5dac372280e5810d8008e54f70b1a5c40bde) to [mistral-inference](https://github.com/mistralai/mistral-inference) (the reference implementation) may be of interest to those working on implementation in llama.cpp.

**Comment by CrispStrobe** [comment]
is it really not supported? did you check with taking the tokenizer etc from Nemo, working with the consolidated safetensors, modify the config.json to something like
```

{
  "architectures": [
    "MistralForCausalLM"
  ],
  "attention_dropout": 0.0,
  "bos_token_id": 1,
  "eos_token_id": 2,
  "head_dim": 128,
  "hidden_act": "silu",
  "hidden_size": 4096,
  "initializer_range": 0.02,
  "intermediate_size": 12288,
  "max_position_embeddings": 32768,
  "model_type": "mistral",
  "num_attention_heads": 32,
  "num_hidden_layers": 36,
  "num_key_value_heads": 8,
  "rms_norm_eps": 1e-05,
  "rope_theta": 100000000.0,
  "sliding_window": null,
  "tie_word_embeddings": false,
  "torch_dtype": "bfloat16",
  "transformers_version": "4.43.0.dev0",
  "use_cache": true,
  "vocab_size": 131072
}
```
? it looks like the established tekken v3 and common architecture?

**Comment by bartowski1182** [comment]
it's likely that it won't work properly until the sliding window changes are added

**Comment by nonetrix** [comment]
Seems to sorta work for me, although I need to apply the prompt template seemingly lol
![image](https://github.com/user-attachments/assets/b5213074-2e3b-4d37-ab64-d9354635adf2)

**Comment by bartowski1182** [comment]
someone on lmstudio discord showed it kind of going off the rails after 2k context

**Comment by CrispStrobe** [comment]
ah i see now, there are others meanwhile, how did you generate your quants, i.e., what guesses at the jsons? (btw i only tried with ctx 2048 yet)

**Comment by bartowski1182** [comment]
I used the conversion done by prince canuma here:
https://huggingface.co/prince-canuma/Ministral-8B-Instruct-2410-HF

The conversion seems to like up with responses from the space hosting the original weights

My guess is that my GGUF conversion would not do long context well but should be fine otherwise

**Comment by CrispStrobe** [comment]
i see. other than long context and seemingly hard to mitigate hallucination (really astonishing imo nowadays!?), it seems to run okay...ish? (my approach was similar, i just manually adjusted the jsons - interestingly, with mostly the same results as prince-canuma, who seemingly used a script for that - and minimally tweaked the hf to gguf conversion script to accept consolidated safetensors too)

**Comment by steampunque** [comment]
> I used the conversion done by prince canuma here: https://huggingface.co/prince-canuma/Ministral-8B-Instruct-2410-HF
> 
> And pulled the tokenizer.json and tokenizer_config.json from Nemo since they matched
> 
> The conversion seems to line up with responses from the space hosting the original weights
> 
> My guess is that my GGUF conversion would not do long context well but should be fine otherwise

Thanks for pointer to prince-canuma.  I grabbed his HF repo, cloned tokenizer.json, tokenizer_config.json, and special_tokens_map.json from Nemo instruct 12B and made a Q6_K quant.

KV memory efficiency looks good, >32K GPU KV available with full offload of Q6_K on a 12G VRAM GPU is good.

Basic functionality looks good, it seems pretty smart:
```
lm "I have 10 apples. I find 3 gold coins in the bottom of a river. The river runs near a big city that has something to do with what I can spend the coins on. I theMinistral-8B-Instruct-2410.IQ4_XS.ggufn lose 4 apples but gain a gold coin. Three birds run into my path and drop 6 apples each. I play an online game and win 6 gold coins but I have to share them equally with my 2 teammates. I buy apples for all the coins I have. The price of an apple is 0.5 coins. How many apples do I have? And where is the river? Use step-by-step reasoning to solve this problem."
```
Let's break down the problem step-by-step:

1. **Initial Apples and Coins:**
   - You start with 10 apples.
   - You find 3 gold coins in the river.

2. **Apples and Coins After Losing and Gaining:**
   - You lose 4 apples: 10 - 4 = 6 apples.
   - You gain 1 gold coin: 3 + 1 = 4 gold coins.

3. **Apples and Coins After Birds Drop Apples:**
   - Three birds drop 6 apples each: 3 birds * 6 apples/bird = 18 apples.
   - Total apples now: 6 (from step 2) + 18 (from birds) = 24 apples.

4. **Apples and Coins After Winning Online Game:**
   - You win 6 gold coins in the online game.
   - You have to share these 6 gold coins equally with your 2 teammates: 6 coins / 3 people = 2 coins per person.
   - Total gold coins now: 4 (from step 2) + 2 (your share) = 6 gold coins.

5. **Apples and Coins After Buying Apples:**
   - The price of one apple is 0.5 coins.
   - You can buy apples with all your 6 gold coins: 6 coins * 2 apples/coin = 12 apples.

6. **Final Count of Apples:**
   - You already have 24 apples from step 3.
   - You buy 12 more apples with your coins.
   - Total apples now: 24 + 12 = 36 apples.Ministral-8B-Instruct-2410.IQ4_XS.gguf

So, you have 36 apples in total.

Regarding the river's location, the problem doesn't provide specific details about where the river is. It only mentions that the river runs near a big city.
```
lm "Q: In the following sentence, explain the antecedent of the pronoun (which thing the pronoun refers to), or state that it is ambiguous:
Sentence: The reporter and the chef will discuss their favorite dishes.
Options:
(A) They will discuss the reporter's favorite dishes
(B) They will discuss the chef's favorite dishes
(C) Ambiguous
A: Let's think step by step,"
```
The sentence "reporter and chef will discuss their favorite dishes" is ambiguous because it is unclear whether "their" refers to the reporter's favorite dishes or the chef's favorite dishes. The pronoun "their" does not have a clear antecedent in the sentence, as it could refer to either the reporter or the chef.

Therefore, the correct answer is (C) Ambiguous.

Quick benches seems on par with other models
Test                 |    Ministral-8B-Instruct-2410 Q6_K  | Mistral 7B Instruct v0.3  Q8_0  | Meta-Llama-3.1-8B-Instruct Q6_K   | gemma-2-9b-it.Q6_KMinistral-8B-Instruct-2410.IQ4_XS.gguf
-------------------|------------------------------------------------|----------------------------------|---------|---------
WINOGRANDE   |  0.748                                                 |      0.751                           |  0.741 |  0.762
LAMBADA       |  0.776                                                  |     0.766                            | 0.747 |  0.735
HELLASWAG      |  0.835             |  0.591        |  0.696             |    0.775       

Strong hellaswag outclasses gemma 2 9b significantly, suggests reasoning will be strong.

No long context tests done until its supported.

**Comment by steampunque** [comment]
> Thank you @steampunque , this is valuable information! Looks like the HF format is in a pull request at the moment: https://huggingface.co/mistralai/Ministral-8B-Instruct-2410/discussions/7

Thanks for heads up, will re-convert when it becomes available.  Though I think prince canumas weights were already good.  I believe this model was overfit on code which erased quite a bit of knowledge and abilities typically present in other SOTA 8B class models.

**Comment by arch-btw** [comment]
> The PR will not convert with convert_hf_to_gguf.py due to a format change in the bpe merges in tokenizer.config. The new format uses an array to contain each merge pair:

Hmm interesting, the pr converted for me without changes, are you sure it's not using the consolidated.safetensors? I'm converting to f32 maybe that's another possible reason.

But yeah, I agree with your view on this model. I've also noticed it's ridiculously sensitive to errors in the prompt format.

**Comment by steampunque** [comment]
> > The PR will not convert with convert_hf_to_gguf.py due to a format change in the bpe merges in tokenizer.config. The new format uses an array to contain each merge pair:
> 
> Hmm interesting, the pr converted for me without changes, are you sure it's not using the consolidated.safetensors? I'm converting to f32 maybe that's another possible reason.
> 
> But yeah, I agree with your view on this model. I've also noticed it's ridiculously sensitive to errors in the prompt format.
> 
> Edit: they also updated a bunch of files in the pr by the way.

convert script is most likely ok but some other python library needs update on my system for the new format, not sure which one (it wasnt the torch library, I updated that).  I have the latest updates, checked sha256s and they match.  Would be a great agent model I believe if the new attention scheme will ever supported otherwise its essentially useless as stands.
